channeldb/payment_control.go

```diff
@@ -538,6 +546,15 @@ func (p *PaymentControl) Fail(paymentHash lntypes.Hash,
 			return err
 		}
 
+		if !p.cfg.KeepFailedPayments &&
+			payment.FailureReason != nil &&
+			payment.Status == StatusFailed {
```

Comment:
The payment may still have some active shards, so it's quite likely that calling `deletePaymentTx` will fail here (see `routing/payment_lifecycle.go`)

Reply:
I took a look at `routing/payment_lifecycle.go` as you suggested. However, `fetchPayment` in `channeldb/payments.go`, which constructs the MPPayment instance, ensures that payment.Status will only be set to StatusFailed if there are no active shards for the payment. I also realized that the `payment.FailureReason != nil` check I had there was redundant since the Fail function always sets the FailureReason to a non-nil value.

Reply:
Resolving this, since I have changed the PR to the point where this conversation is no longer relevant.